File: 46.全排列.ts
/*
 * @lc app=leetcode.cn id=46 lang=typescript
 *
 * [46] 全排列
 */

// @lc code=start


function swap<T>(arr: T[], index1: number, index2: number) {
  const tempElement = arr[index1];
  arr[index1] = arr[index2];
  arr[index2] = tempElement;
  return arr;
}

let count: number = 0;

const backtrack46 = (n: number, output: number[], res: number[][], first: number) => {
  if (first === n) {
      res.push([...output]);
  }
  for (let i = first; i < n; i++) {
      swap(output, first, i);
      backtrack(n, output, res, first + 1);
      swap(output, first, i);
  }
}

function permute(nums: number[]): number[][] {
  const res: number[][] = [];
  const output: number[] = [];

  for (let num of nums) {
      output.push(num);
  }

  const n = nums.length;
  backtrack(n, output, res, 0);
  return res;
};

// @lc code=end

File: leetcode_go/pkg/algorithm/46.全排列.go
package algorithm

/*
 * @lc app=leetcode.cn id=46 lang=golang
 *
 * [46] 全排列
 */

// @lc code=start

func back46(nums *[]int, res *[][]int, first int) {
	n := len(*nums)
	if first == n {
		temp := []int{}
		temp = append(temp, *nums...)
		*res = append(*res, temp)
	}
	for i := first; i < n; i++ {
		(*nums)[first], (*nums)[i] = (*nums)[i], (*nums)[first]
		back46(nums, res, first+1)
		(*nums)[first], (*nums)[i] = (*nums)[i], (*nums)[first]
	}
}

func permute(nums []int) [][]int {
	res := [][]int{}
	back46(&nums, &res, 0)
	return res
}

// @lc code=end

func Permute(nums []int) [][]int {
	return permute(nums)
}
